File: backend/evaluar.py
"""Ejecuta el benchmark médico contra la API real y genera un reporte JSON.

Uso:
  .venv/bin/python evaluar.py --dataset evals/alzheimer.jsonl --dry-run
  .venv/bin/python evaluar.py --dataset evals/alzheimer.jsonl --token TOKEN
  RAG_PASSWORD=... .venv/bin/python evaluar.py --dataset ... --login correo@airobotix.net
  .venv/bin/python evaluar.py --dataset evals/alzheimer.jsonl --repeticiones 3

Repeticiones: la misma pregunta corrida 5 veces dio fidelidad entre 0.33 y
1.00, entre 6 y 10 hops y entre 5 y 15 afirmaciones (2026-09-03). Con una sola
pasada, cualquier "mejoró" o "empeoró" tras un cambio es ruido, así que
`--repeticiones N` corre cada caso N veces y el reporte trae, por caso, el
agregado (`score`/`result`: mediana de lo numérico, mayoría de lo booleano y
la dispersión, ver `app.evaluation.aggregate_runs`) MÁS las N corridas crudas
en `runs`, para poder ir a mirar la corrida que se salió. El resumen y el gate
de release se calculan sobre el agregado: un caso cuenta una vez, no N.

Formato del reporte (schema_version 2):
  {"summary": ..., "repetitions": N, "interrupted": null | {...},
   "cases": [{"definition", "score", "result", "runs": [{"result", "score"}, ...]}]}

`score` es el agregado del caso y con N = 1 coincide con el `score_case` de la
única corrida salvo por las claves aditivas (`runs`, `passed_rate`,
`dispersion`, `found_rate`). `result`, en cambio, NO es la corrida: es un
agregado mínimo con `id`, `question`, `mode`, `runs`, los `errors` vistos y
`metrics`/`dispersion` de coste y latencia (lo único que leen `summarize` y el
tope de coste). La respuesta, las fuentes, los hops y el error de cada corrida
viven enteros en `cases[i].runs[j].result`, también con N = 1.

`interrupted` no es null cuando el reporte es PARCIAL: el bucle se cayó (o se
interrumpió con Ctrl-C) y se escribió lo ya medido en vez de tirarlo; entonces
el gate no se abre aunque lo evaluado pase.
"""
from __future__ import annotations

import argparse
import datetime as dt
import json
import os
import sys
from pathlib import Path
from typing import Any

import httpx

from app.evaluation import EvalCase, aggregate_runs, load_cases, score_case, summarize
from preguntar import obtener_token


def run_case(case: EvalCase, base: str, token: str | None) -> dict[str, Any]:
    headers = {"Authorization": f"Bearer {token}"} if token else {}
    answer: list[str] = []
    sources: list[dict] = []
    hops: list[dict] = []
    metrics: dict = {}
    error: str | None = None

    with httpx.stream(
        "POST",
        f"{base.rstrip('/')}/api/chat",
        json={"session_id": None, "message": case.question, "modo": case.mode},
        headers=headers,
        timeout=330,
    ) as response:
        response.raise_for_status()
        event = ""
        for line in response.iter_lines():
            if line.startswith("event:"):
                event = line.split(":", 1)[1].strip()
                continue
            if not line.startswith("data:"):
                continue
            try:
                data = json.loads(line.split(":", 1)[1].strip())
            except json.JSONDecodeError:
                continue
            if event == "token":
                answer.append(str(data.get("text") or ""))
            elif event == "sources":
                sources = data.get("sources") or []
            elif event == "hop":
                hops.append(data)
            elif event == "metrics":
                metrics = data
            elif event == "error":
                error = str(data.get("detail") or "error desconocido")

    return {
        "id": case.id,
        "question": case.question,
        "mode": case.mode,
        "answer": "".join(answer),
        "sources": sources,
        "hops": hops,
        "metrics": metrics,
        "error": error,
    }


def _run_once(case: EvalCase, base: str, token: str | None) -> dict[str, Any]:
    """Una corrida del caso; CUALQUIER fallo se registra como resultado con
    `error`, no como excepción, para que puntúe (y falle) como las demás.

    Antes solo se atrapaba `httpx.HTTPError`, y cualquier otra cosa (un evento
    SSE con una forma inesperada -> KeyError o TypeError) tumbaba `main()` sin
    escribir el reporte: se perdían todas las corridas ya hechas y ya pagadas,
    que es justo lo que el tope por corrida existe para no desperdiciar. El
    nombre de la excepción va dentro del mensaje porque `str(KeyError("text"))`
    es solo `'text'` y no se entiende sin él.
    """
    try:
        return run_case(case, base, token)
    except Exception as exc:  # noqa: BLE001
        return {
            "id": case.id,
            "question": case.question,
            "mode": case.mode,
            "answer": "",
            "sources": [],
            "hops": [],
            "metrics": {},
            "error": f"{type(exc).__name__}: {exc}",
        }


def main() -> int:
    parser = argparse.ArgumentParser(description="Evalúa fidelidad y recuperación del RAG")
    parser.add_argument("--dataset", type=Path, required=True)
    parser.add_argument("--base", default="http://localhost:8000")
    parser.add_argument("--token")
    parser.add_argument("--login", metavar="CORREO")
    parser.add_argument("--output", type=Path)
    parser.add_argument("--max-usd", type=float)
    # Default 1 para no cambiar lo que mide una corrida normal (ni lo que
    # cuesta): con 1 el agregado es la propia corrida.
    parser.add_argument(
        "--repeticiones",
        type=int,
        default=1,
        metavar="N",
        help="corridas por caso; el reporte agrega (mediana/mayoría) y conserva las N crudas",
    )
    parser.add_argument("--dry-run", action="store_true", help="solo valida el dataset")
    args = parser.parse_args()
    if args.repeticiones < 1:
        parser.error("--repeticiones debe ser >= 1")

    try:
        cases = load_cases(args.dataset)
    except (OSError, ValueError) as exc:
        parser.error(str(exc))
    print(f"Benchmark válido: {len(cases)} casos en {args.dataset}")
    if args.dry_run:
        return 0

    token = args.token
    if args.login:
        password = os.environ.get("RAG_PASSWORD")
        if not password:
            parser.error("define RAG_PASSWORD para usar --login")
        token = obtener_token(args.login, password)

    # Por caso: (definición, score agregado, result agregado, corridas crudas).
    evaluated: list[tuple[EvalCase, dict[str, Any], dict[str, Any], list[dict[str, Any]]]] = []
    spent = 0.0
    # Interrupción imprevista del bucle. No se propaga: el reporte se escribe
    # con las corridas que ya se hicieron (y ya se pagaron) y el código de
    # salida es distinto de 0. Perderlas todas por un fallo en el caso 9 de 10
    # es exactamente lo que el tope por corrida intenta evitar.
    interrupted: BaseException | None = None
    try:
        budget_hit = False
        for index, case in enumerate(cases, start=1):
            runs: list[dict[str, Any]] = []
            for repetition in range(1, args.repeticiones + 1):
                # El tope se mira antes de CADA corrida, no solo antes de cada
                # caso: con N repeticiones un caso multiplica el gasto por N.
                if args.max_usd is not None and spent >= args.max_usd:
                    budget_hit = True
                    break
                label = f"[{index}/{len(cases)}] {case.id}"
                if args.repeticiones > 1:
                    label += f" ({repetition}/{args.repeticiones})"
                print(f"{label}: {case.question}", flush=True)
                result = _run_once(case, args.base, token)
                score = score_case(case, result)
                runs.append({"result": result, "score": score})
                spent += float((result.get("metrics") or {}).get("cost_usd") or 0)
                print("  PASS" if score["passed"] else "  FAIL: " + "; ".join(score["failures"]))
            if not runs:
                print(f"Tope de coste alcanzado antes del caso {case.id}")
                break
            # Si el tope cortó a medias, se agrega lo que hay (k < N corridas)
            # en vez de tirar el gasto ya hecho; `score["runs"]` dice cuántas.
            score, result = aggregate_runs(
                [run["score"] for run in runs], [run["result"] for run in runs]
            )
            evaluated.append((case, score, result, runs))
            if args.repeticiones > 1:
                passed_runs = round(score["passed_rate"] * score["runs"])
                verdict = "PASS" if score["passed"] else "FAIL"
                line = f"  => {verdict} {passed_runs}/{score['runs']} corridas"
                if score["failures"]:
                    line += ": " + "; ".join(score["failures"])
                print(line)
            if budget_hit:
                print(
                    f"Tope de coste alcanzado: {case.id} se agregó con "
                    f"{len(runs)} de {args.repeticiones} corridas"
                )
                break
    except (Exception, KeyboardInterrupt) as exc:  # noqa: BLE001
        # KeyboardInterrupt incluido a propósito: un Ctrl-C a mitad de un
        # benchmark de una hora tampoco debe tirar lo ya medido.
        interrupted = exc
        print(
            f"Benchmark interrumpido tras {len(evaluated)} de {len(cases)} casos: "
            f"{type(exc).__name__}: {exc}",
            file=sys.stderr,
            flush=True,
        )

    if not evaluated:
        # Sin ningún caso agregado no hay nada que resumir: `summarize([])`
        # levantaba ValueError('no hay resultados que resumir') y el proceso
        # moría con un traceback y sin reporte. Pasa con `--max-usd 0` (tope
        # agotado antes de la primera corrida) o si el bucle se cae en el
        # primer caso.
        if interrupted is not None:
            motivo = f"interrumpido ({type(interrupted).__name__})"
        elif args.max_usd is not None:
            motivo = "tope de coste agotado antes de la primera corrida"
        else:
            motivo = "el bucle no llegó a agregar ningún caso"
        print(f"No se evaluó ningún caso ({motivo}): no hay nada que reportar.")
        return 1

    scored = [score for _, score, _, _ in evaluated]
    results = [result for _, _, result, _ in evaluated]
    summary = summarize(scored, results)
    report = {
        "schema_version": 2,
        "created_at": dt.datetime.now(dt.timezone.utc).isoformat(),
        "dataset": str(args.dataset),
        "repetitions": args.repeticiones,
        # Reporte parcial: se escribió igual, pero dice por qué le faltan casos
        # para que nadie lo compare con uno completo creyéndolo completo.
        "interrupted": (
            None
            if interrupted is None
            else {
                "error": f"{type(interrupted).__name__}: {interrupted}",
                "cases_evaluated": len(evaluated),
                "cases_total": len(cases),
            }
        ),
        "summary": summary,
        "cases": [
            {"definition": case.model_dump(), "score": score, "result": result, "runs": runs}
            for case, score, result, runs in evaluated
        ],
    }
    output = args.output or Path("evals/results") / (
        dt.datetime.now().strftime("%Y%m%d-%H%M%S") + ".json"
    )
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(report, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps(summary, ensure_ascii=False, indent=2))
    print(f"Reporte: {output}")
    # Un benchmark interrumpido no abre el gate ni aunque lo medido pasara: no
    # se comprobaron todos los casos.
    return 0 if summary["release_gate_passed"] and interrupted is None else 1


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/evaluar.ts
// Corre el benchmark médico contra el despliegue de Convex y escribe un
// reporte JSON. Port de `backend/evaluar.py` sobre el arnés interno
// (`pruebas:prepararPregunta` + `pruebas:leerTurno`); la puntuación vive en
// `convex/evaluacion/puntuar.ts` y es la misma que medía el Python.
//
// Ejecutar desde frontend/ (lee VITE_CONVEX_URL y CONVEX_DEPLOY_KEY del
// entorno o de .env.local):
//
//   npx vite-node scripts/evaluar.ts --dataset evals/alzheimer.jsonl --dry-run
//   npx vite-node scripts/evaluar.ts --dataset evals/alzheimer.jsonl --correo medica@alzheimerproject.com
//   npx vite-node scripts/evaluar.ts --dataset ... --correo ... --repeticiones 3 --max-usd 2
//
// `--correo` es la cuenta CUYO CORPUS se pregunta: cada persona tiene el suyo,
// así que la misma pregunta desde dos cuentas busca en índices distintos. Las
// conversaciones que crea el benchmark se borran al terminar cada caso
// (`--conservar` las deja para mirarlas en la interfaz).
//
// Repeticiones: la misma pregunta corrida 5 veces dio fidelidad entre 0.33 y
// 1.00 (2026-09-03). Con una sola pasada cualquier "mejoró" es ruido, así que
// `--repeticiones N` corre cada caso N veces y el reporte agrega (mediana de lo
// numérico, mayoría de lo booleano, la dispersión) MÁS las N corridas crudas.
//
// El reporte se escribe SIEMPRE, también si el bucle se interrumpe (Ctrl-C o un
// fallo): lo ya medido y pagado no se tira, y `interrupted` dice que es parcial
// para que nadie lo compare con uno completo. El código de salida es 0 solo si
// el gate de release pasa y el benchmark se completó.
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { ConvexHttpClient } from "convex/browser";
import { makeFunctionReference } from "convex/server";
import {
  agregarCorridas,
  cargarCasos,
  puntuarCaso,
  resumir,
  type Caso,
  type Puntuacion,
  type Resultado,
  type ResultadoAgregado,
} from "../convex/evaluacion/puntuar";

// ---------------------------------------------------------------------------
// Argumentos
// ---------------------------------------------------------------------------
interface Args {
  dataset: string;
  correo: string | null;
  repeticiones: number;
  output: string | null;
  maxUsd: number | null;
  dryRun: boolean;
  conservar: boolean;
  esperaMaxS: number;
}

function leerArgs(argv: string[]): Args {
  const a: Args = { dataset: "", correo: null, repeticiones: 1, output: null, maxUsd: null, dryRun: false, conservar: false, esperaMaxS: 660 };
  for (let i = 0; i < argv.length; i++) {
    const k = argv[i];
    const valor = () => {
      const v = argv[++i];
      if (v === undefined) throw new Error(`falta el valor de ${k}`);
      return v;
    };
    if (k === "--dataset") a.dataset = valor();
    else if (k === "--correo") a.correo = valor();
    else if (k === "--repeticiones") a.repeticiones = Number(valor());
    else if (k === "--output") a.output = valor();
    else if (k === "--max-usd") a.maxUsd = Number(valor());
    else if (k === "--espera-max") a.esperaMaxS = Number(valor());
    else if (k === "--dry-run") a.dryRun = true;
    else if (k === "--conservar") a.conservar = true;
    else throw new Error(`argumento desconocido: ${k}`);
  }
  if (!a.dataset) throw new Error("hace falta --dataset");
  if (!Number.isInteger(a.repeticiones) || a.repeticiones < 1) throw new Error("--repeticiones debe ser un entero >= 1");
  if (a.maxUsd !== null && !(a.maxUsd >= 0)) throw new Error("--max-usd debe ser un número >= 0");
  return a;
}

/** Variables del entorno, con `.env.local` de respaldo (sin dependencias). */
function variable(nombre: string): string | undefined {
  if (process.env[nombre]) return process.env[nombre];
  const ruta = resolve(".env.local");
  if (!existsSync(ruta)) return undefined;
  for (const linea of readFileSync(ruta, "utf8").split(/\r?\n/)) {
    const m = /^\s*([A-Z0-9_]+)\s*=\s*(.*)\s*$/.exec(linea);
    if (m && m[1] === nombre) return m[2].replace(/^["']|["']$/g, "");
  }
  return undefined;
}

// ---------------------------------------------------------------------------
// Convex
// ---------------------------------------------------------------------------
const prepararPregunta = makeFunctionReference<"mutation", { texto: string; modo: string; correo?: string }, { sessionId: string; messageId: string }>(
  "pruebas:prepararPregunta",
);
const leerTurno = makeFunctionReference<"query", { messageId: string }, Turno | null>("pruebas:leerTurno");
const borrarSesion = makeFunctionReference<"mutation", { sessionId: string }, { estado: string }>("pruebas:borrarSesionDePrueba");

interface Turno {
  estado: string | null;
  error: string | null;
  content: string;
  sources: unknown[];
  hops: unknown[];
  plan: unknown[];
  verificacion: unknown;
  metrics: Record<string, unknown>;
}

const FINALES = new Set(["listo", "error", "cancelado"]);

function dormir(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

/** Una corrida del caso: lanza la pregunta y espera el turno final. CUALQUIER
 *  fallo se devuelve como resultado con `error`, no como excepción, para que
 *  puntúe (y falle) como las demás y el reporte no pierda lo ya medido. */
async function correrCaso(cliente: ConvexHttpClient, caso: Caso, args: Args): Promise<{ resultado: Resultado; sessionId: string | null }> {
  let sessionId: string | null = null;
  try {
    const { sessionId: sid, messageId } = await cliente.mutation(prepararPregunta, {
      texto: caso.question,
      modo: caso.mode,
      ...(args.correo ? { correo: args.correo } : {}),
    });
    sessionId = sid;
    const limite = Date.now() + args.esperaMaxS * 1000;
    let turno: Turno | null = null;
    while (Date.now() < limite) {
      await dormir(3000);
      turno = await cliente.query(leerTurno, { messageId });
      if (turno === null) throw new Error("el mensaje desapareció mientras se esperaba");
      if (turno.estado && FINALES.has(turno.estado)) break;
    }
    if (!turno || !turno.estado || !FINALES.has(turno.estado)) {
      throw new Error(`el turno no terminó en ${args.esperaMaxS} s (estado: ${turno?.estado ?? "?"})`);
    }
    const metrics = { ...turno.metrics };
    // El verificador deja su informe en la fila (`verificacion`) y su resumen
    // en la telemetría (`metrics.meta.verificacion`), que es lo que lee el
    // puntuador; si la telemetría no lo trajera, se reconstruye de la fila.
    const meta = (metrics.meta ?? {}) as Record<string, unknown>;
    if (!meta.verificacion && turno.verificacion && typeof turno.verificacion === "object") {
      const v = turno.verificacion as { fidelidad?: number | null; afirmaciones?: Array<{ veredicto?: string }> };
      const cuenta = (x: string) => (v.afirmaciones ?? []).filter((a) => a.veredicto === x).length;
      metrics.meta = { ...meta, verificacion: { fidelidad: v.fidelidad ?? null, no_sostenidas: cuenta("no_sostenida"), sin_verificar: cuenta("sin_verificar") } };
    }
    return {
      sessionId,
      resultado: {
        id: caso.id,
        question: caso.question,
        mode: caso.mode,
        answer: turno.content,
        sources: turno.sources,
        hops: turno.hops,
        metrics,
        error: turno.estado === "error" ? (turno.error ?? "error sin detalle") : turno.estado === "cancelado" ? "turno cancelado" : null,
      },
    };
  } catch (exc) {
    return {
      sessionId,
      resultado: {
        id: caso.id, question: caso.question, mode: caso.mode, answer: "", sources: [], hops: [], metrics: {},
        error: `${exc instanceof Error ? exc.constructor.name : "Error"}: ${exc instanceof Error ? exc.message : String(exc)}`,
      },
    };
  }
}

// ---------------------------------------------------------------------------
// Principal
// ---------------------------------------------------------------------------
async function main(): Promise<number> {
  const args = leerArgs(process.argv.slice(2));
  const casos = cargarCasos(readFileSync(args.dataset, "utf8"), args.dataset);
  console.log(`Benchmark válido: ${casos.length} casos en ${args.dataset}`);
  if (args.dryRun) return 0;
  if (!args.correo) throw new Error("hace falta --correo: la cuenta cuyo corpus se pregunta");

  const url = variable("VITE_CONVEX_URL");
  const clave = variable("CONVEX_DEPLOY_KEY");
  if (!url || !clave) throw new Error("hacen falta VITE_CONVEX_URL y CONVEX_DEPLOY_KEY (entorno o .env.local)");
  const cliente = new ConvexHttpClient(url);
  cliente.setAdminAuth(clave);

  type Evaluado = { caso: Caso; score: Puntuacion; result: ResultadoAgregado; runs: Array<{ result: Resultado; score: Puntuacion }> };
  const evaluados: Evaluado[] = [];
  let gastado = 0;
  let interrumpido: unknown = null;
  let cortado = false;
  process.on("SIGINT", () => {
    cortado = true;
    console.error("\nInterrumpido: se escribirá lo ya medido.");
  });

  try {
    let topeAlcanzado = false;
    for (const [indice, caso] of casos.entries()) {
      if (cortado) throw new Error("interrumpido por la usuaria");
      const runs: Array<{ result: Resultado; score: Puntuacion }> = [];
      for (let rep = 1; rep <= args.repeticiones; rep++) {
        if (args.maxUsd !== null && gastado >= args.maxUsd) {
          topeAlcanzado = true;
          break;
        }
        if (cortado) throw new Error("interrumpido por la usuaria");
        const etiqueta = `[${indice + 1}/${casos.length}] ${caso.id}` + (args.repeticiones > 1 ? ` (${rep}/${args.repeticiones})` : "");
        console.log(`${etiqueta}: ${caso.question}`);
        const { resultado, sessionId } = await correrCaso(cliente, caso, args);
        const score = puntuarCaso(caso, resultado);
        runs.push({ result: resultado, score });
        gastado += Number((resultado.metrics ?? {}).cost_usd ?? 0) || 0;
        console.log(score.passed ? "  PASS" : "  FAIL: " + score.failures.join("; "));
        if (sessionId && !args.conservar) {
          try {
            await cliente.mutation(borrarSesion, { sessionId });
          } catch (exc) {
            console.warn(`  (no se pudo borrar la conversación de prueba: ${exc instanceof Error ? exc.message : String(exc)})`);
          }
        }
      }
      if (!runs.length) {
        console.log(`Tope de coste alcanzado antes del caso ${caso.id}`);
        break;
      }
      const [score, result] = agregarCorridas(runs.map((r) => r.score), runs.map((r) => r.result));
      evaluados.push({ caso, score, result, runs });
      if (args.repeticiones > 1) {
        const aprobadas = Math.round((score.passed_rate ?? 0) * (score.runs ?? 1));
        let linea = `  => ${score.passed ? "PASS" : "FAIL"} ${aprobadas}/${score.runs} corridas`;
        if (score.failures.length) linea += ": " + score.failures.join("; ");
        console.log(linea);
      }
      if (topeAlcanzado) {
        console.log(`Tope de coste alcanzado: ${caso.id} se agregó con ${runs.length} de ${args.repeticiones} corridas`);
        break;
      }
    }
  } catch (exc) {
    interrumpido = exc;
    console.error(`Benchmark interrumpido tras ${evaluados.length} de ${casos.length} casos: ${exc instanceof Error ? exc.message : String(exc)}`);
  }

  if (!evaluados.length) {
    console.log("No se evaluó ningún caso: no hay nada que reportar.");
    return 1;
  }
  const summary = resumir(evaluados.map((e) => e.score), evaluados.map((e) => e.result));
  const reporte = {
    schema_version: 3,
    created_at: new Date().toISOString(),
    dataset: args.dataset,
    correo: args.correo,
    repetitions: args.repeticiones,
    interrupted:
      interrumpido === null
        ? null
        : { error: interrumpido instanceof Error ? interrumpido.message : String(interrumpido), cases_evaluated: evaluados.length, cases_total: casos.length },
    summary,
    cases: evaluados.map((e) => ({ definition: e.caso, score: e.score, result: e.result, runs: e.runs })),
  };
  const salida = args.output ?? resolve("evals/results", `${new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19)}.json`);
  mkdirSync(dirname(salida), { recursive: true });
  writeFileSync(salida, JSON.stringify(reporte, null, 2), "utf8");
  console.log(JSON.stringify(summary, null, 2));
  console.log(`Reporte: ${salida}`);
  return summary.release_gate_passed && interrumpido === null ? 0 : 1;
}

process.exitCode = await main();
